part1

In [28]:
# import the required libraries
import pandas as pd 
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import make_column_transformer
from sklearn.compose import ColumnTransformer
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier 
from sklearn.metrics import accuracy_score 

In [29]:
# Import the dataset
df = pd.read_csv("/kaggle/input/hotel-reservations/Hotel Reservations.csv", index_col="Booking_ID")

In [30]:
# check the top 5 rows of the data
df.head()

no_of_adults  no_of_children  no_of_weekend_nights  \
Booking_ID                                                       
INN00001               2               0                     1   
INN00002               2               0                     2   
INN00003               1               0                     2   
INN00004               2               0                     0   
INN00005               2               0                     1   

            no_of_week_nights type_of_meal_plan  required_car_parking_space  \
Booking_ID                                                                    
INN00001                    2       Meal Plan 1                           0   
INN00002                    3      Not Selected                           0   
INN00003                    1       Meal Plan 1                           0   
INN00004                    2       Meal Plan 1                           0   
INN00005                    1      Not Selected                           0   

           room_type_reserved  lead_time  arrival_year  arrival_month  \
Booking_ID                                                              
INN00001          Room_Type 1        224          2017             10   
INN00002          Room_Type 1          5          2018             11   
INN00003          Room_Type 1          1          2018              2   
INN00004          Room_Type 1        211          2018              5   
INN00005          Room_Type 1         48          2018              4   

            arrival_date market_segment_type  repeated_guest  \
Booking_ID                                                     
INN00001               2             Offline               0   
INN00002               6              Online               0   
INN00003              28              Online               0   
INN00004              20              Online               0   
INN00005              11              Online               0   

            no_of_previous_cancellations  \
Booking_ID                                 
INN00001                               0   
INN00002                               0   
INN00003                               0   
INN00004                               0   
INN00005                               0   

            no_of_previous_bookings_not_canceled  avg_price_per_room  \
Booking_ID                                                             
INN00001                                       0               65.00   
INN00002                                       0              106.68   
INN00003                                       0               60.00   
INN00004                                       0              100.00   
INN00005                                       0               94.50   

            no_of_special_requests booking_status  
Booking_ID                                         
INN00001                         0   Not_Canceled  
INN00002                         1   Not_Canceled  
INN00003                         0       Canceled  
INN00004                         0       Canceled  
INN00005                         0       Canceled

In [31]:
# check the shape of the data
df.shape
#samples-features

(36275, 18)

In [32]:
# check the missing values
df.isna().sum()

no_of_adults                            0
no_of_children                          0
no_of_weekend_nights                    0
no_of_week_nights                       0
type_of_meal_plan                       0
required_car_parking_space              0
room_type_reserved                      0
lead_time                               0
arrival_year                            0
arrival_month                           0
arrival_date                            0
market_segment_type                     0
repeated_guest                          0
no_of_previous_cancellations            0
no_of_previous_bookings_not_canceled    0
avg_price_per_room                      0
no_of_special_requests                  0
booking_status                          0
dtype: int64

In [33]:
# check the datatype of the columns
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 36275 entries, INN00001 to INN36275
Data columns (total 18 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   no_of_adults                          36275 non-null  int64  
 1   no_of_children                        36275 non-null  int64  
 2   no_of_weekend_nights                  36275 non-null  int64  
 3   no_of_week_nights                     36275 non-null  int64  
 4   type_of_meal_plan                     36275 non-null  object 
 5   required_car_parking_space            36275 non-null  int64  
 6   room_type_reserved                    36275 non-null  object 
 7   lead_time                             36275 non-null  int64  
 8   arrival_year                          36275 non-null  int64  
 9   arrival_month                         36275 non-null  int64  
 10  arrival_date                          36275 non-null  int64  
 11  market_seg

In [34]:
# check the statistical summary of the data
df.describe()

no_of_adults  no_of_children  no_of_weekend_nights  no_of_week_nights  \
count  36275.000000    36275.000000          36275.000000       36275.000000   
mean       1.844962        0.105279              0.810724           2.204300   
std        0.518715        0.402648              0.870644           1.410905   
min        0.000000        0.000000              0.000000           0.000000   
25%        2.000000        0.000000              0.000000           1.000000   
50%        2.000000        0.000000              1.000000           2.000000   
75%        2.000000        0.000000              2.000000           3.000000   
max        4.000000       10.000000              7.000000          17.000000   

       required_car_parking_space     lead_time  arrival_year  arrival_month  \
count                36275.000000  36275.000000  36275.000000   36275.000000   
mean                     0.030986     85.232557   2017.820427       7.423653   
std                      0.173281     85.930817      0.383836       3.069894   
min                      0.000000      0.000000   2017.000000       1.000000   
25%                      0.000000     17.000000   2018.000000       5.000000   
50%                      0.000000     57.000000   2018.000000       8.000000   
75%                      0.000000    126.000000   2018.000000      10.000000   
max                      1.000000    443.000000   2018.000000      12.000000   

       arrival_date  repeated_guest  no_of_previous_cancellations  \
count  36275.000000    36275.000000                  36275.000000   
mean      15.596995        0.025637                      0.023349   
std        8.740447        0.158053                      0.368331   
min        1.000000        0.000000                      0.000000   
25%        8.000000        0.000000                      0.000000   
50%       16.000000        0.000000                      0.000000   
75%       23.000000        0.000000                      0.000000   
max       31.000000        1.000000                     13.000000   

       no_of_previous_bookings_not_canceled  avg_price_per_room  \
count                          36275.000000        36275.000000   
mean                               0.153411          103.423539   
std                                1.754171           35.089424   
min                                0.000000            0.000000   
25%                                0.000000           80.300000   
50%                                0.000000           99.450000   
75%                                0.000000          120.000000   
max                               58.000000          540.000000   

       no_of_special_requests  
count            36275.000000  
mean                 0.619655  
std                  0.786236  
min                  0.000000  
25%                  0.000000  
50%                  0.000000  
75%                  1.000000  
max                  5.000000

In [35]:
# Print the number of samples
print('number of samples is:', len(df))
# Print the number of features
print('number of features is:', len(df.columns))

number of samples is: 36275
number of features is: 18


In [36]:
# Check the number of categorical features
categorical_features = df.select_dtypes(include=['object']).columns
print("number of categorical features is:", len(categorical_features))

number of categorical features is: 4


In [37]:
# Check the distribution of the target label
print('Distribution of the target variableis :')
df['booking_status'].value_counts()

Distribution of the target variableis :


Not_Canceled    24390
Canceled        11885
Name: booking_status, dtype: int64

Shape of dataset is : (119390, 31)
Number of samples: 36275
Number of features: 18
Number of categorical features: 4
Distribution of target variableis:
Not_Canceled    24390
Canceled        11885

In [38]:
# Split the dataset into X and y
X = df.drop('booking_status', axis=1)
y = df['booking_status']

In [39]:
# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=100)

In [40]:
# Create a decision tree classifier and fit it to the training data
model = DecisionTreeClassifier(random_state=100)
model.fit(X_train, y_train)

# Make predictions on the testing data
y_pred = model.predict(X_test)

# Evaluate the accuracy of the model
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy is:", accuracy)

ValueError: could not convert string to float: 'Meal Plan 1'

- There is an error due to categorical columns in the dataset
- we will usr use numeric datatypes insted

In [41]:
# Select only the numeric features
X_numeric = X.select_dtypes(exclude=['object'])

In [43]:
# Split the dataset into training and testing sets
X_train_numeric, X_test_numeric, y_train, y_test = train_test_split(X_numeric, y, test_size=0.2, random_state=100)

# Create a decision tree classifier and fit it to the training data
clf_numeric = DecisionTreeClassifier(random_state=100)
clf_numeric.fit(X_train_numeric, y_train)

# Make predictions on the testing data
y_pred_numeric = clf_numeric.predict(X_test_numeric)

# Evaluate the accuracy of the model
accuracy_numeric = accuracy_score(y_test, y_pred_numeric)
print("Accuracy (numeric features only):", accuracy_numeric)

Accuracy (numeric features only): 0.8581667815299793


This Accuracy is %85 accurate which is good

8

In [44]:
from sklearn.model_selection import GridSearchCV


# Define the hyperparameter grid
param_grid = {
    'criterion': ['gini', 'entropy'],
    'max_depth': [5, 10, 15, None],
    'max_features': ['sqrt', 'log2', None],
    'splitter': ['best', 'random']
}

# Create a decision tree classifier
clf_gs = DecisionTreeClassifier(random_state=100)

# Use GridSearchCV to find the best hyperparameters
clf_gs = GridSearchCV(clf_gs, param_grid, cv=5)
clf_gs.fit(X_train_numeric, y_train)

# Print the best hyperparameters
print("Best hyperparameters:", clf_gs.best_params_)

Best hyperparameters: {'criterion': 'gini', 'max_depth': 15, 'max_features': None, 'splitter': 'best'}


In [45]:
# Create a decision tree classifier with the best hyperparameters
clf_best = DecisionTreeClassifier(criterion ='gini', max_depth = 15, max_features =None, splitter ='best',random_state=100)
clf_best.fit(X_train_numeric, y_train)

# Make predictions on the testing data
y_pred_best = clf_best.predict(X_test_numeric)

# Evaluate the accuracy of the model
accuracy_best = accuracy_score(y_test, y_pred_best)
print("Accuracy (numeric features only, with best hyperparameters):", accuracy_best)

Accuracy (numeric features only, with best hyperparameters): 0.8583046175051688


In [46]:
print("Accuracy (numeric features only, without tuning hyperparameters):", accuracy_numeric)
print("Accuracy (numeric features only, with best hyperparameters):", accuracy_best)

Accuracy (numeric features only, without tuning hyperparameters): 0.8581667815299793
Accuracy (numeric features only, with best hyperparameters): 0.8583046175051688


- Accuracy is almost same . improved a litlle 

part2

In [47]:
column_transformer = make_column_transformer((OneHotEncoder(), ['type_of_meal_plan', 'room_type_reserved', 'market_segment_type']),remainder='passthrough')

# First, we must transform the entire dataset
X_transformed = column_transformer.fit_transform(X)
X_transformed = pd.DataFrame(data=X_transformed)

#second,we divide the transformed dataset into train/test dataset
X_train, X_test, y_train, y_test = train_test_split(X_transformed, y, random_state = 100)

In [48]:
# Encode categorical features
ct = ColumnTransformer(transformers=[('encoder', OneHotEncoder(), ['type_of_meal_plan', 'room_type_reserved', 'market_segment_type'])], remainder='passthrough')
X = ct.fit_transform(df.drop('booking_status', axis=1))

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=100)

# Create a decision tree classifier
clf_all = DecisionTreeClassifier(random_state=100)

# Fit the classifier to the training data
clf_all.fit(X_train, y_train)

# Make predictions on the testing data
y_pred_all = clf_all.predict(X_test)

# Evaluate the model
acc_all = accuracy_score(y_test, y_pred_all)
print("Accuracy (all features):", acc_all)

Accuracy (all features): 0.8691936595451413


This model is %86.91 acuurate which means is better than the numeric only model and almost as good as the best-hyperparamter model